### PR TITLE
always enable grpc stats filter

### PR DESCRIPTION
**Pull request by ramaraochavali**
- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
lgtm, @kyessenov can you double check this?

**Comment by costinm** [comment]
Why ? How does it impact a non-grpc protocol ? Any perf impact ?  If someone declares a protocol as http2 - will they
now higher memory and some stats they don't expect/want ? 

Not very familiar with the grpc stats filter - not sure where is the best place to put this info, I suspect there is a 
lot of confusion around  the gRPC stats generated by Envoy, our own Istio telemetry - the new OpenTelemetry
spec and its semantic conventions - plus the grpc.io defined metrics. All that applies to gRPC - but there is also
'Connect' variant that the filter supports ( nice), plus otel seems to have generic 'rpc' metrics independent of 
protocol. 

Putting this all together - we have a huge mess of metrics we generate, some by default and some opt-in. 

IMO if we have one metric enabled by default ( including in the proposed 'Istio safe'), it is the OTel one. 
Official gRPC ( including proxyless ) still recommends grpc.io metrics, so it may be ok to emit them as well.
But Envoy gRPC stats are envoy-only and don't match either of the options used in a 'non-envoy-proxy' case.

My preference is to still keep them as an option - I would even add a second 'disable envoy stats' so
Istio safe can turn off stats that we don't consider part of Istio stable metrics surface.

**Comment by kyessenov** [comment]
@costinm The impact is limited to gRPC (as detected by content-type) but it does imply a performance hit as the data buffers are now scanned for frame boundaries. 

grpc_stats filter needs to evolve to produce gRPC message Otel telemetry - if there an associated envoy issue to bring it to Otel standards?

**Comment by kyessenov** [comment]
@costinm I believe the intent of the PR is a workaround for multiple routes on the same HCM. I agree about the perf impact, but this is not a new feature - we had this filter enabled for "protocol: grpc" for a very long time. So there's no backwards compatibility risk here, it's now detecting gRPC on the data path instead of the less reliable xDS path.

For the future changes - we need to start a discussion in Envoy to align grpc_stats with gRPC Otel. Don't think we need to block on that here.

**Comment by costinm** [comment]
On Fri, Mar 24, 2023 at 10:06 AM Kuat ***@***.***> wrote:

> @costinm <https://github.com/costinm> I believe the intent of the PR is a
> workaround for multiple routes on the same HCM. I agree about the perf
> impact, but this is not a new feature - we had this filter enabled for
> "protocol: grpc" for a very long time. So there's no backwards
> compatibility risk here, it's now detecting gRPC on the data path instead
> of the less reliable xDS path.
>
It is a backward incompatibility risk - users who declare the port as http2
would now see lower perf and have no way to turn it off.

I agree we can't remove the behavior if protocol=gRPC - for same backward
compat reasons, in the normal install.
But we can remove it in the 'safe Istio' which we are planning - it is not
a documented or tested/officially supported behavior.

I have no problem with having options to have it on for all protocols
including http ( with the new 'buf' gRPC impl it may
also be for HTTP/1.1 ), but also to turn it off and only have the official
telemetry in safe mode to make sure 'safe istio'
users don't rely on undocumented/unsupported features.



> For the future changes - we need to start a discussion in Envoy to align
> grpc_stats with gRPC Otel. Don't think we need to block on that here.
>

Agreed, don't want to block it - just to make sure it is an option that
users can opt-in for ( in the case of using it for protocol != gRPC)
and opt out ( for protocol  = gRPC ). Maybe a
"GRPC_STATS_FILTER=all|grpc|none" or something like this ?

File: pilot/pkg/networking/core/v1alpha3/listener_builder.go [hunk 1]
```diff
@@ -386,9 +386,7 @@ func (lb *ListenerBuilder) buildHTTPConnectionManager(httpOpts *httpListenerOpts
 		filters = append(filters, xdsfilters.GrpcWeb)
 	}
 
-	if httpOpts.protocol.IsGRPC() {
-		filters = append(filters, xdsfilters.GrpcStats)
-	}
+	filters = append(filters, xdsfilters.GrpcStats)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/listener_builder.go:389]: If we always enable grpc stats, why nor grpcweb as well?

**Review by hzxuzhonghu — COMMENTED** [review]
I think we should check the protocol as well to be clean. Otherwise it is confusing with grpc filter for pure HTTP services

**Comment by costinm** [comment]
On Fri, Mar 24, 2023 at 6:53 PM Zhonghu Xu ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pilot/pkg/networking/core/v1alpha3/listener_builder.go
> <https://github.com/istio/istio/pull/44094#discussion_r1148249537>:
>
> > @@ -386,9 +386,7 @@ func (lb *ListenerBuilder) buildHTTPConnectionManager(httpOpts *httpListenerOpts
>  		filters = append(filters, xdsfilters.GrpcWeb)
>  	}
>
> -	if httpOpts.protocol.IsGRPC() {
> -		filters = append(filters, xdsfilters.GrpcStats)
> -	}
> +	filters = append(filters, xdsfilters.GrpcStats)
>
> If we always enable grpc stats, why nor grpcweb as well?
>

I think it's reasonable to enable grpcweb if protocol is GRPC.

File: pilot/pkg/networking/core/v1alpha3/listener_builder.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/listener_builder.go:389]: Sure we can do. But let us do it separately.

**Comment by ramaraochavali** [comment]
> @costinm The impact is limited to gRPC (as detected by content-type) but it does imply a performance hit as the data buffers are now scanned for frame boundaries.

@kyessenov I thought it is just based on content type and if content type is not gRPC we do not do any thing in this filter. Where re data buffers scanned if it is not gRPC?

If enabling by default is a concern, I can add a flag `ALWAYS_ENABLE_GRPC_STATS_FILTER` and it we enable it if the protocol is gRPC or if the flag is set.

**Comment by ramaraochavali** [comment]
> I think we should check the protocol as well to be clean. Otherwise it is confusing with grpc filter for pure HTTP services

That is what we do today and this adds but nothing is done for pure HTTP services. The filter will be skipped

**Comment by costinm** [comment]
On Fri, Mar 24, 2023 at 10:30 PM Rama Chavali ***@***.***>
wrote:

> @costinm <https://github.com/costinm> The impact is limited to gRPC (as
> detected by content-type) but it does imply a performance hit as the data
> buffers are now scanned for frame boundaries.
>
> @kyessenov <https://github.com/kyessenov> I thought it is just based on
> content type and if content type is not gRPC we do not do any thing in this
> filter. Where re data buffers scanned if it is not gRPC?
>
> If enabling by default is a concern, I can add a flag
> ALWAYS_ENABLE_GRPC_STATS_FILTER and it we enable it if the protocol is
> gRPC or if the flag is set.
>
> It is not scanned if content type is not gRPC. But the user may set
appProtocol to http2 instead of grpc - and the filter would not be added.
With this change - grpc traffic that was treated as http2 and didn't get
scanned will be slowed down.

I'm not saying it's the correct behavior to rely on a user incorrectly
setting protocol as http2 when it really is grpc - but it can still be a
regression.

If we were to promote "envoy stats as a supported telemetry integration" to
supported API - it would be treated like the other telemetry
providers, and we would use the new telemetry API to let the user pick
which workloads generate this. However IMHO we should try to
move towards OpenTelemetry defined metric names instead of envoy-only.
Maybe we could still use the stats filter and translate the
metrics in agent or otel collector.

I think just adding the env var is good enough for now. If we really want
to support envoy-specific stats - it would be a larger change and
probably impact bootstrap generator ( where we filter a lot of stats ) and
other parts of the code.

Can you add more context - are the 'official' gRPC metrics not sufficient ?
Do you need the streaming messages count ?

**Comment by ramaraochavali** [comment]
By official you mean istio telemetry we don't use it. We rely on envoy
stats for actual grpc related stats, we need this filter. Anyway i have
added the flag. Please check.

**Comment by costinm** [comment]
On Sat, Mar 25, 2023 at 10:06 PM Rama Chavali ***@***.***>
wrote:

> By official you mean istio telemetry we don't use it. We rely on envoy
> stats for actual grpc related stats, we need this filter. Anyway i have
> added the flag. Please check.
>
>
It is very reasonable to not use istio telemetry, and I think the PR looks
good with the new flag.

But I would appreciate few small changes (optional):
- instead of bool - make it "ENVOY_[GRPC_]STATS" - "on", "off" ( with
default empty or "compat" ).
- if off - disable it even if protocol is grpc.

I would use "ENVOY_STATS" instead of ENVOY_GRPC_STATS - and if it is set to
'on' document that
it will generate envoy stats for grpc - but also possibly other components.
This could be used for example
to un-mask some of the stats in bootstrap, or to enable other stats module
and options without having
to introduce an env var for each of them.

For "off" - I think it is useful for users who chose istio telemetry - or
in future OTel telemetry. I don't know why
we enabled it in the first place - but with ambient and 'istio safe' we
have an opportunity to fix few things.

**Comment by costinm** [comment]
In other words - instead of one setting for each protocol ( in global env
variables ) - we should make it simpler
for users to use envoy stats if they want.

We could take a step further and move this to telemetry API - for example
by defining an internal envoy-stats
provider, which would allow users to use a first-class API to enable envoy
stats at workload/namespace or istiod
level with dynamic changes supported. It is more complicated - but if you (
and others ) are using envoy stats
it seems a much better API.

In an environment where other mesh or gateway implementations based on
envoy are used - it is likely best
to enable this option and treat it as all other telemetry providers.

**Comment by ramaraochavali** [comment]
Those are good points but need more thought on whether we need a special flag for envoy telemetry or not. AFAIK, there are no other stats specific filters in Envoy so we do not do any thing special on the xds path for using envoy telemetry except for grpc_stats.  @hzxuzhonghu @howardjohn @kyessenov WDYT?

**Comment by hzxuzhonghu** [comment]
Personally, i do not like adding more and more env/flags, both are anti-UX patterns to me.

**Comment by kyessenov** [comment]
I'd be fine with a special "revert" flag in case it is really a performance issue for someone (we're guessing right now), but not with a generic `ENVOY_STATS` flag. I don't think we need another API to change Envoy stats.

**Comment by ramaraochavali** [comment]
> I'd be fine with a special "revert" flag in case it is really a performance issue for someone (we're guessing right now), but not with a generic ENVOY_STATS flag. I don't think we need another API to change Envoy stats.

Do you think we should default the new flag `ALWAYS_ENABLE_GRPC_STATS` to true so that people can revert if needed (if they see perf issues)?

**Comment by howardjohn** [comment]
What exactly does this do?

We have https://istio.io/latest/docs/reference/config/metrics/ `istio_request_messages_total` message as part of core Istio metrics. However, I am never seeing this incremented when I send gRPC traffic, even with this PR.

**Comment by kyessenov** [comment]
@howardjohn That feature got broken by me accidentally - since noone complained, I assume noone cares :)

Rama is not using Istio telemetry, he only wants direct Envoy telemetry.

@ramaraochavali I'm fine with default-on. If I remember measuring it a long time ago, the cost was fairly small (<5% CPU).

**Comment by howardjohn** [comment]
When was it broken?

**Comment by costinm** [comment]
I completely agree with avoiding more env/flags, in particular when we have
proper APIs for this.
But at the same time if something is risky and is not part of a
supported/core feature, and is globally enabled we need some controls.
I don't think 'on by default' for this is right either - I think it was
wrong to have the stats filter enabled by default for grpc
with no way for the user to control it. Istio telemetry has plenty of
problems - but it is what we document and is on by default,
using envoy stats should be an option but not default.

I am ok with the PR as is now if we can't agree to use the telemetry API.
There are other XDS paths involving envoy telemetry - see the bootstrap
file where a lot of stats are disabled (AFAIK). I don't know if there are
other filters related to envoy stats - but if they get added in the future
they
could be part of the same 'use envoy telemetry instead of istio telemetry'
option.

**Comment by ramaraochavali** [comment]
@kyessenov  just for my learning want to get your thoughts on this questions - 

@kyessenov I thought it is just based on content type and if content type is not gRPC we do not do any thing in this filter. Where re data buffers scanned if it is not gRPC?

@costinm I think we have discussing about a assumed perf problem - but not completely validated.  Ideally it should be ON by default because that is what gRPC stats do today (and it is internal to Istio on how build HCM)  and people can turn off if they can not tolerate the perf hit. Having said that, I am fine to turn off by default as long as we have ability for people to enable it (especially at gateways).

**Comment by kyessenov** [comment]
@ramaraochavali It's not doing anything for non-gRPC. The discrepancy is because in the prior the filter was inserted for "grpc" ports, and this PR changes it to use content-type. It was possible to have gRPC on "h2" port before, and the filter was not inserted.

**Comment by costinm** [comment]
Yes, also we already made the mistake of turning features on by default a few times ( mixer and the policy checks ? ).

Unless a feature is documented and marked as beta or stable - I don't think it should be on by default for the entire cluster.

I am very supportive of using envoy stats or even promoting it at some point, but IMO we shouldn't have 2 telemetry options on by default at the same time.

We should discuss in the WG our stance on OpenTelemetry - I would be very happy if Envoy had native support for the metrics defined in the 'semantic conventions' or if collector could translate from envoy names to the common ones. That may be a strong argument to make it default in ambient or the 'safe istio' mode.

**Comment by kyessenov** [comment]
@costinm Unfortunately, Kiali depends on the istio_requests_messages_count metric (https://github.com/istio/istio/issues/44144), so we don't have much leeway - it needs to be supported since we don't have an alternative.

**Review by howardjohn — APPROVED** [review]
I think this should be always on. Its a requirement for a core stable API.

It has nothing to do with "using envoy metrics" or "OTEL"; this is a (part of a) simple bug fix for broken telemetry (https://github.com/istio/istio/issues/44144).

Discussions about evolving telemetry are welcome but orthogonal. Can we remove the ALWAYS_ENABLE_GRPC_STATS_FILTER flag and make this always on

**Comment by costinm** [comment]
Why would the envoy stats filter for gRPC be on ? Is it listed as a
supported feature of Istio ? Does it have tests, docs, etc ?

Don't we have a process for making changes that add features on by default
?

**Comment by costinm** [comment]
Wait - you mean that istio_request_message_count depends on this envoy
filter, and if we enable this filter it'll start working ?

**Comment by howardjohn** [comment]
This filter is an implementation detail of a requirement for **core Istio metrics API** https://istio.io/latest/docs/reference/config/metrics/. Its not just envoy metrics

**Comment by costinm** [comment]
Ok, if you mean istio metrics depend on this filter - I have no problem
with having it on by default.

I didn't realize Istio metrics depend on the grpc filter.

**Comment by howardjohn** [comment]
I did not understand it either until today :slightly_smiling_face:

**Comment by zirain** [comment]
> I did not understand it either until today 🙂

+1

**Comment by ramaraochavali** [comment]
Good to know so many people (including me) learnt a thing with this PR :wink: . I have removed the flag

Adding envoy stats as another provider seems like a good idea. @kyessenov might more details around Otel stuff in Envoy. Let us discuss that separately.

**Comment by johnma14** [comment]
Does this need to be back ported to 1.17?

**Comment by ramaraochavali** [comment]
> Does this need to be back ported to 1.17?

I think so. I have added the label.

**Pull request merged**